File: cpp/67a_1867A_GreenGoldDogArrayAndPermutation.py
def solve(a, n):
    b = list(a)
    b.sort(reverse=True)
    hmap = {}
    for i in range(n):
        hmap[b[i]] = i+1
    ret = []
    for i in range(n):
        ret.append(hmap[a[i]])
        hmap[a[i]] -= 1
    return ret

def main():
    ans = []
    t = int(input())
    for _ in range(t):
        n = int(input())
        a = list(map(int, input().split()))
        ans.append(solve(a, n))
    for i in ans:
        for j in i:
            print(j, end = " ")
        print()

if __name__ == "__main__":
    main()

File: cpp/67a_1867A_GreenGoldDogArrayAndPermutation.cpp
#include<bits/stdc++.h>

using namespace std;

vector<int> solve(vector<int> a, int n){
    vector<int> b(a.begin(), a.end());
    sort(b.rbegin(), b.rend());
    unordered_map<int, int> hmap;
    vector<int> ans;
    for(int i = 0; i < n; i++){
        hmap[b[i]] = i+1;
    }
    for(int i = 0; i < n; i++){
        ans.push_back(hmap[a[i]]);
        hmap[a[i]]--;
    }
    return ans;
}

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);

    int t, n, temp;
    vector<vector<int>> ans;
    cin>>t;

    for(int i = 0; i < t; i++){
        cin>>n;
        vector<int> a;
        for(int j = 0; j < n; j++){
            cin>>temp;
            a.push_back(temp);
        }
        ans.push_back(solve(a, n));
    }

    for(auto i : ans){
        for(auto j : i){
            cout<<j<<" ";
        }
        cout<<"\n";
    }
    return 0;
}